### PR TITLE
darwin.stdenv: improvements and overrideSDK rewrite

### PR DESCRIPTION
**Pull request by reckenrode**
## Description of changes

This PR makes several improvements to the Darwin stdenv situation.

### Adds SDK root

It adds the SDK version to the stdenv as a new `darwin.apple_sdk.sdkRoot` package. This is done via a hook that adds `-isysroot` with a path to a stub SDK that just defines version information (based on the `xcbuild.sdk`). This is needed because clang passes `-platform_version <deployment target> <sdk version>` unconditionally to the linker, and the SDK version is inferred to be the same as the deployment target if `-isysroot` or `SDKROOT` do not point to SDK from which it can get the version.
 
macOS implements different behaviors depending on the SDK version. There are two notable examples: dark mode and version checks.

macOS requires the linked SDK to be at least 10.14 for dark mode to work. Currently, x86_64-darwin binaries in nixpkgs do not work with dark mode. With the SDK root specified, clang will pass the correct SDK version, ensuring that binaries linked against the 11.0 SDK work as expected.

For version checks, macOS returns a compatible version when the linked SDK is older than 11.0. I ran into that issue while working on https://github.com/NixOS/nixpkgs/issues/236414. MetalD3D checks the macOS version, but because the load commands in Wine’s binary indicate an older SDK, macOS returns a “compatible” version: 10.20. Since the check is for 14, the check fails, and MetalD3D does not work as expected.

Closes #265139.

### Fleshes out `overrideSDK`

This PR includes a rewrite of `overrideSDK`, addressing most of the issues identified in https://github.com/NixOS/nixpkgs/issues/242666#issuecomment-1803866452. This was necessary to support building Rust applications such as lapce, neovide, and wezterm on x86_64-darwin using the 11.0 SDK. It is also necessary to fix building applications that use `wrapGAppsHook` and the 11.0 SDK.

According to ofborg’s evalulation performance check, the rewritten `overrideSDK` performs about the same as the previous version even though it is more capable: it supports replacing inputs of all types, supports additional SDK-based overrides (libobjc, libs, private frameworks), and should be easier to maintain. It should also work with cross-compilation, but this is untested due to needing https://github.com/NixOS/nixpkgs/pull/256590 to fix cross-compilation support.

I plan to address overriding the SDK for Rust in a separate PR. There are some changes to `rustPlatform` needed to make it work robustly with overriding the SDK. Documentation will also follow in a separate PR.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by reckenrode** [comment]
@toonn

**Comment by reckenrode** [comment]
The force push added support for multiple outputs in the override (though I don’t believe any of the packages currently overridden require it).

**Review by toonn — COMMENTED** [review]
Everything looks good so far. I'll look at override-sdk.nix in another review. Didn't want to have you wait on the rest.

File: pkgs/os-specific/darwin/apple-sdk/sdkRoot.nix
Comment (reckenrode) [reply pkgs/os-specific/darwin/apple-sdk/sdkRoot.nix:44]: I wasn’t sure either. I copied it from cc-wrapper. I can drop its

Comment (reckenrode) [reply pkgs/os-specific/darwin/apple-sdk/sdkRoot.nix]: Dropped. I pushed an update that reforms `override-sdk.nix` according to the new RFC and adds support for compilers to the mapping. If it has a libc, the libc will get swapped to the requested SDK’s libc.

**Comment by reckenrode** [comment]
Aside from adding compiler support, I added an optimization to reduce evaluation time by only building the mapping once per unique package set instead of doing it for every one then deduping at the end. I’m curious about the ofborg performance evaluation because the previous one was *terrible*.

**Comment by reckenrode** [comment]
I managed to improve the terrible performance by ~30%, but the increase in evaluation time is still terrible. Need to look into some stuff on optimizing nix expressions without losing the generality of this solution.

**Comment by reckenrode** [comment]
I fixed the merge conflict. I also optimized the algorithm. I left it as a separate commit, so the non-optimized version can serve as a reference. Running the ofborg check locally, it’s only a 4.6% regression (from 386s to 404s). That’s a _lot_ better considering it was as high as 600% and was 400% after my last attempt at optimization.

What I ended up doing was using `genericClosure` to calculate all dependencies for an input, then I sort and fold over them to calculate the replacements. I then update the original inputs using the new mapping I generate from that.

I’m aware of an issue with Qt 6 where it gets the wrong SDK root. The other overrides appear to work correctly though, so that’s strange. I’m planning to look into that next.

**Comment by reckenrode** [comment]
The Qt 6 issue is because it’s using `apple_sdk_11_0.stdenv` instead of using `overrideSDK`.

**Comment by reckenrode** [comment]
@ofborg eval

**Review by toonn — COMMENTED** [review]
Mostly noticed some typos.

I honestly can't say I've understood everything in detail and I haven't done any testing yet but I wanted to provide at least a first review of overrideSDK.

I wish it was possible to with a single genericClosure invocation rather than the toposort but I think that would require the alternative solution you had in mind with memoization and that does sound worse than the current implementation.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:296]: Treating it as a constraint is needed to make updating SDKs not be a breaking change. Otherwise, updating from the 14.0 to the 14.3 SDK would break everything that used `overrideSDK stdenv "14.0";`. The constraint for a bare number would be similar to a `^` constraint with some special handling for Darwin SDK versioning.

For example:`14.0` would match `14.3` but not `15`; `10.13` would match `10.13.6` but not `10.14`.

(Supporting other types of constraints later is possible, but I don’t see a use case for SDK versioning.)

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:353]: The sets here only have two keys: `key` and one of `pkgs, `replacement`, or `replacementList`. It’s not an issue because the filter should remove those other two. If one of those creeps in, it’s an error, and evaluation should fail. The only reason for the remapping is make everything conform to `listToAttrs`.

One possibility is defining `name` as an alias for `key` and rename `replacement` to `value`. That should let me drop the remapping step. Would that be better?

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:151]: I could do a prepass with `grep`. I was trying to keep things simple.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (toonn) [reply pkgs/stdenv/darwin/override-sdk.nix:296]: The interpretation of a string argument can change, that seems sensible. But I could see a use for `darwinMaxSdkVersion` for example, probably with a name that doesn't imply it means something similar to `darwinMinSdkVersion`.

It's true that the APIs are probably stable enough that it's not very important though.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (toonn) [reply pkgs/stdenv/darwin/override-sdk.nix:353]: No, the current way is fine. I just wasn't certain the attributes were exclusive.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:271]: I don’t think that will work. `toposort` is doing a DFS using the comparison function to define which nodes are adjacent to the current node. I’d need to change `getPropagatedDependencies` to return a mapping of dependency to parents, which doesn’t seem worth doing just to avoid a `reverseList`.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:296]: That could be done as a new constraint (such as `<=14` for the newest SDK prior to the 14.0 SDK). I just don’t want to support it without an actual need for it yet.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:284]: That’s … interesting. I wonder how this is even working or if it’s working. It’s working on the input’s direct dependencies, which is almost certainly wrong. Hmm.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:136]: If I don’t create the output, won’t the build fail?

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:148]: Good catch. It is not guaranteed. ruby-modules puts files in `nix-support/gem-meta`.

**Comment by reckenrode** [comment]
@toonn I think I addressed most of the feedback in the latest commit. I’m refraining from squashing due to the complexity of the changes. Once everything is approved, I can squash the optimization commit down.

The only thing I haven’t addressed is the remapping logic. I need to recursively walk `nix-support` and figure out a solution to the replacement issue.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix]: Would `lib.escapeRegex` be sufficient to avoid that problem?

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix]: I used it in an earlier version of `getPropagatedInputs`, but it’s not needed separately anymore. I’ve removed it from the current version.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix]: I ended up refactoring this to reuse the function since it’s effectively the same.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix]: It took some work, but I was able to make it use `dependencies`. I had to deal with the fact that inputs are not always derivations.

**Comment by reckenrode** [comment]
Setting to draft for now while I figure out why Qt 5 is getting mixed SDK dependencies.

**Comment by reckenrode** [comment]
The Qt 5 issue appears to be that the package built by luck before. It is getting both 10.12 and 11.0 SDK frameworks on master, but the 11.0 SDK frameworks were listed first in `NIX_CFLAGS_COMPILE`, so clang would find 11.0 SDK headers before 10.12 SDK headers. It was using the 10.12 Libsystem, which I’m surprised worked.

**Review by toonn — COMMENTED** [review]
Figured it's more convenient to have the review sooner rather than waiting for you to squash and then you having to wait for the review.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:62]: Remapping is done by `outPath`. `darwin.Libsystem` and `apple_sdk.Libsystem` have the same paths.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:55]: `version` is a set. It’s a workaround for dealing with xcodebuild. It needs some work not to hardcode its SDK root and toolchain stuff, but that’s out of scope for this PR. Once that’s done, `version` and the xcodebuild override can be dropped because it should do the right thing regardless of which SDK it was built against.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:142]: It shouldn’t refer to the original wrapper. `bintools.bintools` is the unwrapped bintools. `bintools.coreutils_bin` and bintools.gnugrep_bin` are the coreutils and gnugrep packages that were passed into the previous, wrapped bintools. I have to rewrap because I can’t use `override` to replace the libc.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:170]: Considering I refer to it as a proxy package elsewhere, I’ll update this to `mkProxyPackage` for consistency.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:182]: It should only ever recurse one level at a time. `$targetPath` will either be `${!outputName}/nix-support` or a subfolder of that.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:207]: I see what you mean now. Yeah, that’s a problem. The original iterated over the source paths, but I dropped that to avoid having to pass them as a file and parse them back out. The same reasoning is what motivated me to try to use sed to do the replacements. I’ll look into a solution for both.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:276]: I’m going to look at other solutions. I might do the replacements with `substitute`, and hopefully the performance won’t be too bad.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:317]: Yeah, that’s redundant. I’ll drop it.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (toonn) [reply pkgs/stdenv/darwin/override-sdk.nix:62]: Even on current x86_64-darwin?

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (toonn) [reply pkgs/stdenv/darwin/override-sdk.nix:55]: And we can't extract the version set from the sdkRoot package?

Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:55]: It’s more that I don’t want to make this bigger than it already is by also making changes to xcodebuild. xcodebuild needs to set the SDK root, but it also creates a toolchain from the clang in its stdenv.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:62]: Yes. I set it in 36202add013a8a81b75b61de4f53955fd91b2900.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:276]: The solution I am pursuing is setting `__structuredAttrs = true;`, which will serialized the attributes as JSON and set them as associative arrays in bash. That lets me avoid parsing the values using `passAsFile`. For replacements, I’m just using `substituteInPlace` iteratively over the `replacements` associative array.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix]: I attempted to reword the comment to better explain what’s happening.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (toonn) [inline pkgs/stdenv/darwin/override-sdk.nix:181]:
```suggestion
      # Take advantage of the fact that replacements and sourceOutputs will be passed
```

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (toonn) [inline pkgs/stdenv/darwin/override-sdk.nix:218]: Would it be worth keeping track of whether any file ever differs from the sourceFile here? In the event that happens we can replace the entire package with a single symlink to the source package.

**Review by toonn — COMMENTED** [review]
One more typo and one more question but IMO this is pretty much done. :+1:

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:181]: Good catch. I originally called the variable `sourceOutPaths` but switched it because `sourceOutputs` read better in the build script.

File: pkgs/stdenv/darwin/override-sdk.nix
Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:218]: `mkProxyPackage` checks whether `replacements` is empty before building the package (and returns the original package if it is). If it’s non-empty, that means there’s at least one input that will be replaces in `nix-support`.

When the `replacePropagatedInputs` bash function recurses, it tries to replace the folder with a symlink if it contains only symlinks. That is untested due to a lack of packages with propagated inputs in a subfolder.

Comment (toonn) [reply pkgs/stdenv/darwin/override-sdk.nix:218]: Ah, so you're checking if the replacements applied to a file, not whether the replacements are actually not a change at all?

Comment (reckenrode) [reply pkgs/stdenv/darwin/override-sdk.nix:218]: When `replacements` is generated in nix, any replacements that match the original are filtered out. If that results in an empty set, then the build is skipped.

By the time the build is being run, it should find at least one replacement. If it doesn’t find any, that would mean a package has propagated inputs that are not recorded in (at least) `$out/nix-support/*propagated*`.

I don’t know how that would happen because recording propagated inputs done automatically for you. The package would have to do something strange on its own to subvert that.

I could try to detect that situation and turn the package into a symlink back to the original, but I hesitate to make that change without having a test case.

**Review by toonn — APPROVED** [review]
LGTM

**Comment by reckenrode** [comment]
@AndersonTorres Sorry about the ping. A Meson I need to build locally patch accidentally snuck into the PR.

**Comment by reckenrode** [comment]
The latest force push rebases on current staging to pick up the libiconv changes.

**Comment by reckenrode** [comment]
@toonn I squashed the wip commits and set the status to ready. I also added a commit dropping unused curl assertions from the stdenv bootstrap. The contents of the PR should otherwise be the same as before the squash.

**Review by toonn — APPROVED** [review]
Still LGTM : )

**Pull request merged**
